src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -428,6 +1072,116 @@ void two_index_constraint(
       constraint, gamma2, spacetime_normal_one_form, inverse_spacetime_metric,
       spacetime_normal_vector, three_index_constraint);
 }
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::a<DataType, SpatialDim, Frame> f_constraint(
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::ia<DataType, SpatialDim, Frame>& d_gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& d_pi,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& d_phi,
+    const Scalar<DataType>& gamma2,
+    const tnsr::iaa<DataType, SpatialDim, Frame>&
+        three_index_constraint) noexcept {
+  auto constraint =
+      make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(pi, 0.0);
+  f_constraint<SpatialDim, Frame, DataType>(
+      &constraint, gauge_function, d_gauge_function, spacetime_normal_one_form,
+      spacetime_normal_vector, inverse_spatial_metric, inverse_spacetime_metric,
+      pi, phi, d_pi, d_phi, gamma2, three_index_constraint);
+  return constraint;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void f_constraint(
+    gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::ia<DataType, SpatialDim, Frame>& d_gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& d_pi,
+    const tnsr::ijaa<DataType, SpatialDim, Frame>& d_phi,
+    const Scalar<DataType>& gamma2,
+    const tnsr::iaa<DataType, SpatialDim, Frame>&
+        three_index_constraint) noexcept {
+  for (auto& component : *constraint) {
```

Comment:
The `constraint` passed in by pointer should have its size checked and set if incorrect. Also, we'd prefer `std::fill` to the for loop.

Reply:
How do I know what the correct size should be?

Reply:
The `constraint` passed by pointer should have its size checked against any of the other tensors passed in. There is a general utility for handling the checking of a size for both fundamental types and vector types (when either could be used, as is the case here for `DataType`). The utility is found in `Utilities/ContainerHelpers.hpp`, and can be called with `get_size(container)`. This will return 1 when the `DataType` is `double` or any other fundamental type.

Reply:
I'm largely fine with this version, any errors will probably be caught quickly. However, what we've been requesting elsewhere is to check the size and if it is wrong to set the variable passed by pointer to a freshly constructed object with an appropriate size. (see for instance the newer versions of #1179 )

Reply:
I’m happy changing it..just thought you’d rather know about and fix the coding errors instead of silently fixing it. But I’ll change it as you prefer!